add global keyboard shortcuts for music controls

cuikho210 opened the pull request:
I am not entirely sure whether the current approach is correct.  
Please let me know if I am doing something wrong, and I will revise it accordingly.

This PR adds several global shortcuts for music controls.  
They are quite important for the desktop experience (at least for me).

## Changes

- Create `Shortcuts` component directory.
- Wrap the main application builder in `lib/main.dart` with `GlobalShortcutManager`.
- Created `platform_helper.dart` to handle platform-specific logic.
- Updated tooltips to display keyboard shortcuts.

## Global Shortcuts

| Shortcut           | Action                 | Description                  |
|--------------------|------------------------|------------------------------|
| Space              | Toggle Playback        | Play / pause                 |
| Ctrl + N           | Skip to Next           | Play next track              |
| Ctrl + P           | Skip to Previous       | Play previous track          |
| Ctrl + →           | Seek Forward           | Seek forward 30 seconds      |
| Ctrl + ←           | Seek Backward          | Seek backward 5 seconds      |
| Ctrl + ↑           | Volume Up              | Increase volume by 5%        |
| Ctrl + ↓           | Volume Down            | Decrease volume by 5%        |
| Ctrl + L           | Toggle Loop Mode       | Switch repeat modes          |
| Ctrl + S           | Toggle Playback Order  | Toggle shuffle / normal      |

## Related Issues

- closes #1370

cuikho210 committed:
feat: add global keyboard shortcuts for music controls

- Create `GlobalShortcutManager`, `MusicControlActions`, and `MusicControlIntents` in a new `Shortcuts` component directory
- Wrap the main application builder in `lib/main.dart` with `GlobalShortcutManager`
- Support shortcuts for play/pause, skip, previous, seek, and volume adjustments
- Ensure shortcuts (like Space for play/pause) are ignored when a text field is focused to prevent input conflicts

flloschy commented:
Not an expert too, but looks good imo
Though im wondering if music_control_intents.dart is really required? I feel like this could make adding new shortcuts a bit convoluted when you need to edit 3 different files just do to one thing.
I didnt look in the docs, so maybe this is required :P
In any case it'd be neat to add a bit of docs to the contribution guidelines :D (of couse this can wait till the end!)


Some more (not thought through) ideas for shortcuts which could be cool:
ALT + L: Toggle Lyrics
ALT + W: Toggle Fullscreen player (kinda like closing a browser tab, thats the idea at least)
ALT + D: Open Downloads
ALT + S: Open Settings
ALT + Q: Open Quick connect menu
A: Accept currently open modal
C: Cancel currently open modal
CTRL + C: Clear Queue
CTRL + P: Play currently visible album/playlist/tracks
CTRL + S: Shuffle currently visible album/playlist/tracks
CTRL + O: Toggle Offline Mode
CTRL + F: Toggle Favourite 

(ALT -> Navigation, CTRL -> Player Controls)

One last thing:
I would find it really cool if shortcuts send a Snackbar message saying what the heck they just did. I often fat finger some shortcuts and am totally lost to what just happend because there is absolutely no feeback

cuikho210 commented:
Thanks for the feedback!

I’ll revisit the current structure and try to make it easier to extend and maintain going forward.

I’ll also add Snackbar feedback for shortcuts that are easy to trigger accidentally or cause confusion, such as seek and next/previous.

For now, this PR is intentionally scoped to music control shortcuts only.  
Navigation-related shortcuts sound useful, but I’ll leave those for a follow-up PR to keep things focused.

flloschy commented:
> I’ll revisit the current structure and try to make it easier to extend and maintain going forward.

Its not too much development burden to add the intents in another file, totally an optional thing to make that simpler. Though if you have the time that would still be cool :p

> For now, this PR is intentionally scoped to music control shortcuts only.

Totally fair! Thanks for tackling this long standing feature request :D

cuikho210 committed:
refactor(shortcuts): consolidate and simplify music control shortcuts

cuikho210 committed:
show snackbar when skipping tracks via shortcuts

cuikho210 committed:
feat(shortcuts): add keyboard shortcut to toggle loop mode

cuikho210 committed:
feat(shortcuts): add keyboard shortcut to toggle playback order

cuikho210 committed:
docs: add instructions for contributing keyboard shortcuts

Komodo5197 commented on lib/components/Shortcuts/music_control_shortcuts.dart:
```diff
@@ -0,0 +1,184 @@
+  };
+}
+
+class _CustomCallbackAction<T extends Intent> extends CallbackAction<T> {
```
Give these two callback actions less generic names.

Komodo5197 commented on lib/components/Shortcuts/global_shortcut_manager.dart:
```diff
@@ -0,0 +1,34 @@
+  Widget build(BuildContext context) {
+    return Shortcuts(
+      shortcuts: {
+        // Music control shortcuts
```
I think with the addition of shortcuts, we need some sort of shortcut/gesture guide in the settings (or maybe somewhere even more obvious?) so people can actually find out about these.

cuikho210 replied:
What do you think about using tooltips on buttons to describe their shortcuts? For elements like the volume bar, we could place a small descriptive label next to them.

Chaphasilor replied:
Adding the shortcuts to the tooltips sounds like a good idea! This should be dynamic of course, in case shortcuts become customizable in the future, and can therefore not depend on the translation strings. We could either add slots for the shortcut to the existing translation strings, or just always put the shortcut in a new line.  

If you find good ways to overlay the shortcuts over things like the volume bar, that would also work I think. Additionally, we could show a snackbar if someone clicks the volume bar (etc.) for the first time, telling them about the shortcut. That would help with discoverability, but is optional for this PR.

cuikho210 replied:
Regarding the volume bar, I think a text label works best. Using an overlay makes it less intuitive and adds unnecessary complexity. What do you think?

Chaphasilor replied:
Showing a snackbar should be simple, but sure, you can also always add the shortcut to both ends of the bar. The contrast is currently a bit problematic, but I'll fix that separately.  
Also, make sure to only add the text labels for the shortcuts on desktop platforms, naturally :)

cuikho210 committed:
rename internal action classes for music controls

cuikho210 committed:
feat(shortcuts): add tooltips and UI hints for keyboard controls

Inform users about available keyboard shortcuts through visual hints and tooltips to improve discoverability.

- Create lib/utils/platform_helper.dart to provide platform-specific key names like ⌘ or Ctrl
- Add shortcut information to tooltips in lib/components/PlayerScreen/player_buttons.dart, player_buttons_loop_mode.dart, and player_buttons_playback_order.dart
- Wrap the progress slider in lib/components/PlayerScreen/control_area.dart with a tooltip explaining seek shortcuts
- Add a volume control instruction text to lib/menus/output_menu.dart
- Define new localization strings for shortcut hints in lib/l10n/app_en.arb

Chaphasilor commented on lib/components/PlayerScreen/player_buttons.dart:
```diff
@@ -96,6 +99,7 @@ class PlayerButtons extends StatelessWidget {
               container: true,
               excludeSemantics: true,
               child: IconButton(
+                tooltip: "$modKeyDisplay+N",
```
these tooltips should be parsed from the actual shortcuts, to ensure they are always in sync

Chaphasilor commented on lib/components/Shortcuts/global_shortcut_manager.dart:
```diff
@@ -0,0 +1,34 @@
+    return Shortcuts(
+      shortcuts: {
+        // Music control shortcuts
+        const SingleActivator(LogicalKeyboardKey.space): const TogglePlaybackIntent(),
```
would it be possible to reverse this map? that should allow assigning multiple shortcuts to a single action (like both ctrl+n and ctrl+alt+→ for skip to next) while keeping things organized.  

I'm guessing the `Shortcuts` widget is some Flutter-internal, so we'd probably have to convert the reversed map into this structure.  
optional for merging, but might be useful?

cuikho210 replied:
Does the use of LogicalKeySet that I just updated in global_shortcut_manager.dart meet your expectations?

Chaphasilor replied:
Hmm, I think it should be a `Map<Intent, List<LogicalKeySet>>` instead of `Map<Intent, LogicalKeySet>`, right? Otherwise, how would you assign multiple keyboard shortcuts to the same action?

Chaphasilor commented on lib/components/Shortcuts/music_control_shortcuts.dart:
```diff
@@ -0,0 +1,184 @@
+import 'package:flutter/material.dart';
+import 'package:get_it/get_it.dart';
+
+class TogglePlaybackIntent extends Intent {
```
Is this `Intent` compatible with Android's intent system?

cuikho210 replied:
No, this Intent is part of Flutter's Shortcuts and Actions system. It's used for mapping keyboard shortcuts to internal app actions and has no connection to the Android native Intent system.

Chaphasilor commented on lib/utils/platform_helper.dart:
```diff
@@ -0,0 +1,9 @@
+import 'dart:io';
+
+class ShortcutKeyDisplay {
+  static String get primaryModifier => Platform.isIOS || Platform.isMacOS ? '⌘' : 'Ctrl';
+  static String get shift => Platform.isIOS || Platform.isMacOS ? '⇧' : 'Shift';
+  static String get alt => Platform.isIOS || Platform.isMacOS ? '⌥' : 'Alt';
+  static String get enter => Platform.isIOS || Platform.isMacOS ? '⏎' : 'Enter';
+  static String get backspace => Platform.isIOS || Platform.isMacOS ? '⌫' : 'Backspace';
+}
```
I'm not sure if all of these characters will render on all devices? Should we be using icons instead?

cuikho210 replied:
Good point. I'll switch to Icons to ensure they render correctly on all devices.

Komodo5197 replied:
Isn't this just Unicode?  I don't see why all devices wouldn't have it.

cuikho210 replied:
I'm not entirely sure about this. I'm actually reconsidering because Icons might be a pain to implement, especially with l10n messages. I still think Icons would look more consistent, but I'm not sure if it's worth the extra complexity. What do you think?

Chaphasilor replied:
It really just depends on if the `⌫` character and such are supported by all fonts, or if proper fallbacks are working now? I just wanted to avoid getting the outline boxes or question mark symbols. But I guess we can run with this approach for now, and see if any users report issues with it. So let's leave it like it is.

Chaphasilor reviewed (commented):
@cuikho210 thanks for the PR! And sorry for "ignoring" it for so long, I've been taking some time off to work on another project.  

Really appreciate the changes, and this already looks pretty mergeable! I left some (mostly optional) comments that you might wanna address, just some finishing touches :)

cuikho210 commented:
@Chaphasilor I'm currently away on a trip. I'll be back in a few days to finalize this PR. Thanks a lot for your review!

affhp commented:
Thank you so much for this, this really bring the desktop experience to the next level.

cuikho210 committed:
feat(shortcuts): centralize definitions and add dynamic display logic

- Create `GlobalShortcuts` class in `global_shortcut_manager.dart` to manage shortcut-to-intent mappings centrally
- Implement `getDisplay` logic to generate human-readable strings from keyboard keys
- Update `control_area.dart`, `player_buttons.dart`, `player_buttons_loop_mode.dart`, and `player_buttons_playback_order.dart` to use dynamic tooltips
- Modify `output_menu.dart` and `app_en.arb` to display dynamic volume and seek hints

cuikho210 commented:
Sorry for the delay. I've just updated the PR. If there are any further comments, let me know and I will address them.

cuikho210 committed:
support multiple key bindings per intent

In `lib/components/Shortcuts/global_shortcut_manager.dart`:
- Change `_raw` map to associate intents with a list of `LogicalKeySet` to allow for multiple triggers.
- Update `shortcutMap` getter to flatten the intent-to-keys mapping during registration.
- Refactor `getDisplay` to use the first available key set for UI representation.

cuikho210 committed:
restrict volume shortcut hints to desktop platforms

cuikho210 committed:
chore(shortcuts): remove debug print in global shortcut manager

yougotwill commented:
@cuikho210 Thank you for working on this! Life happened and I never got around to working on this issue. I appreciate your hard work!

yougotwill commented:
@Chaphasilor May we please get this merged? 🙏

Chaphasilor committed:
fix changing volume, fix shortcuts not responding, improve shortcut hints, add logging

Chaphasilor committed:
Merge remote-tracking branch 'origin/redesign' into pr/cuikho210/1493

Chaphasilor commented:
@cuikho210 @yougotwill sorry for not getting to this earlier. I just gave it a test and made some final fixes. Should all be working now.

Thanks again for tackling this @cuikho210, it's a really nice addition! Definitely helps to add some maturity to the desktop version :)
